src/query/graphite/native/builtin_functions.go

```diff
@@ -242,6 +242,40 @@ func timeShift(
 	}, nil
 }
 
+
+// delay shifts all samples later by an integer number of steps. This can be used
+// for custom derivative calculations, among other things. Note: this will pad
+// the early end of the data with NaN for every step shifted. delay complements
+// other time-displacement functions such as timeShift and timeSlice, in that
+// delay is indifferent about the step intervals being shifted.
+func delay(
+	ctx *common.Context,
+	singlePath singlePathSpec,
+	steps int,
+) (ts.SeriesList, error) {
+		input := ts.SeriesList(singlePath)
+		output := make([]*ts.Series, input.Len())
+
+		for i, series := range input.Values {
+			delayedVals := delayValuesHelper(ctx, *series, steps)
+			delayedSeries := ts.NewSeries(ctx, series.Name(), series.StartTime(), delayedVals)
+			renamedSeries := delayedSeries.RenamedTo(fmt.Sprintf("delay(%s,%d)", delayedSeries.Name(), steps))
+			output[i] = renamedSeries
```

Comment:
nit: Prefer to pre-allocate slice but use append for collecting values.

e.g. 
```
output := make([]*ts.Series, 0, input.Len())
// ...
output = append(output, renamedSeries)
```

We find using this wherever possible causes less bugs in general.